Test test_prometheus_protocols is flaky

https://s3.amazonaws.com/clickhouse-test-reports/71135/c6c66b84e9c235e25ca5b0c6537420f590ba55e1/integration_tests__asan__old_analyzer__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel3_0.log

```
Evaluating query: up
Evaluation time: 1730226644.112345
Requesting http://172.16.2.3:9090/api/v1/query?query=up&time=1730226644.112345

...

>           raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]
E           urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='172.16.2.3', port=9090): Max retries exceeded with url: /api/v1/query?query=up&time=1730226644.112345 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f76c366ae90>: Failed to establish a new connection: [Errno 111] Connection refused'))

/usr/local/lib/python3.10/dist-packages/urllib3/util/retry.py:515: MaxRetryError
```

In server log 

```
2225:2024.10.29 18:31:07.344139 [ 10 ] {} <Debug> PROMETHEUS-Session: 600e28b2-df40-4c4f-b389-a3e8c501b2d0 Authenticating user 'default' from 172.16.2.3:40594
2226:2024.10.29 18:31:07.344226 [ 10 ] {} <Debug> PROMETHEUS-Session: 600e28b2-df40-4c4f-b389-a3e8c501b2d0 Authenticated with global context as user 94309d50-4f52-5250-31bd-74fecac179db
2228:2024.10.29 18:31:07.353860 [ 10 ] {} <Debug> PROMETHEUS-Session: 600e28b2-df40-4c4f-b389-a3e8c501b2d0 Logout, user_id: 94309d50-4f52-5250-31bd-74fecac179db
```

cc @vitlibar

--- a/tests/integration/helpers/cluster.py
+++ b/tests/integration/helpers/cluster.py
@@ -744,11 +744,13 @@ class ClickHouseCluster:
         # available when with_prometheus == True
         self.with_prometheus = False
         self.prometheus_writer_host = "prometheus_writer"
+        self.prometheus_writer_ip = None
         self.prometheus_writer_port = 9090
         self.prometheus_writer_logs_dir = p.abspath(
             p.join(self.instances_dir, "prometheus_writer/logs")
         )
         self.prometheus_reader_host = "prometheus_reader"
+        self.prometheus_reader_ip = None
         self.prometheus_reader_port = 9091
         self.prometheus_reader_logs_dir = p.abspath(
             p.join(self.instances_dir, "prometheus_reader/logs")
@@ -2728,6 +2730,16 @@ class ClickHouseCluster:
 
         raise Exception("Can't wait LDAP to start")
 
+    def wait_prometheus_to_start(self):
+        self.prometheus_reader_ip = self.get_instance_ip(self.prometheus_reader_host)
+        self.prometheus_writer_ip = self.get_instance_ip(self.prometheus_writer_host)
+        self.wait_for_url(
+            f"http://{self.prometheus_reader_ip}:{self.prometheus_reader_port}/api/v1/query?query=time()"
+        )
+        self.wait_for_url(
+            f"http://{self.prometheus_writer_ip}:{self.prometheus_writer_port}/api/v1/query?query=time()"
+        )
+
     def start(self):
         pytest_xdist_logging_to_separate_files.setup()
         logging.info("Running tests in {}".format(self.base_path))
@@ -3083,11 +3095,22 @@ class ClickHouseCluster:
                     f"http://{self.jdbc_bridge_ip}:{self.jdbc_bridge_port}/ping"
                 )
 
-            if self.with_prometheus:
+            if self.with_prometheus and self.base_prometheus_cmd:
                 os.makedirs(self.prometheus_writer_logs_dir)
                 os.chmod(self.prometheus_writer_logs_dir, stat.S_IRWXU | stat.S_IRWXO)
                 os.makedirs(self.prometheus_reader_logs_dir)
                 os.chmod(self.prometheus_reader_logs_dir, stat.S_IRWXU | stat.S_IRWXO)
+
+                prometheus_start_cmd = self.base_prometheus_cmd + common_opts
+
+                logging.info(
+                    "Trying to create Prometheus instances by command %s",
+                    " ".join(map(str, prometheus_start_cmd)),
+                )
+                run_and_check(prometheus_start_cmd)
+                self.up_called = True
+                logging.info("Trying to connect to Prometheus...")
+                self.wait_prometheus_to_start()
 
             clickhouse_start_cmd = self.base_cmd + ["up", "-d", "--no-recreate"]
             logging.debug(

--- a/tests/integration/test_prometheus_protocols/test.py
+++ b/tests/integration/test_prometheus_protocols/test.py
@@ -20,7 +20,7 @@ node = cluster.add_instance(
 
 def execute_query_on_prometheus_writer(query, timestamp):
     return execute_query_impl(
-        cluster.get_instance_ip(cluster.prometheus_writer_host),
+        cluster.prometheus_writer_ip,
         cluster.prometheus_writer_port,
         "/api/v1/query",
         query,
@@ -30,7 +30,7 @@ def execute_query_on_prometheus_writer(query, timestamp):
 
 def execute_query_on_prometheus_reader(query, timestamp):
     return execute_query_impl(
-        cluster.get_instance_ip(cluster.prometheus_reader_host),
+        cluster.prometheus_reader_ip,
         cluster.prometheus_reader_port,
         "/api/v1/query",
         query,

